src/node.cc

```diff
@@ -1303,86 +1303,187 @@ void AddPromiseHook(v8::Isolate* isolate, promise_hook_func fn, void* arg) {
-                               Local<Value> recv,
-                               const Local<Function> callback,
-                               int argc,
-                               Local<Value> argv[],
-                               async_context asyncContext) {
-  // If you hit this assertion, you forgot to enter the v8::Context first.
-  CHECK_EQ(env->context(), env->isolate()->GetCurrentContext());
+  v8::TryCatch real_try_catch;
+};
+}
 
-  Local<Object> object;
+class InternalCallbackScope {
+ public:
+  InternalCallbackScope(Environment* env,
+                        v8::Local<v8::Value> object,
+                        const async_context& asyncContext);
+  ~InternalCallbackScope();
+  void Close();
 
-  Environment::AsyncCallbackScope callback_scope(env);
-  bool disposed_domain = false;
+  inline bool Failed() const { return failed_; }
+  inline void MarkAsFailed() { failed_ = true; }
+  inline v8::Local<v8::Value> GetFailureReturnValue() const {
+    return failure_return_value_;
+  }
+  inline bool IsInnerMakeCallback() const {
+    return callback_scope_.in_makecallback();
+  }
 
-  if (recv->IsObject()) {
-    object = recv.As<Object>();
+  inline void UseTryCatch() {
+    CHECK_EQ(try_catch_, nullptr);
+    try_catch_ = new _TryCatch(env_->isolate());
+    try_catch_->real_try_catch.SetVerbose(true);
   }
 
+ private:
+  Environment* env_;
+  async_context async_context_;
+  v8::Local<v8::Object> object_;
+  Environment::AsyncCallbackScope callback_scope_;
+  bool failed_ = false;
+  bool pushed_ids_ = false;
+  bool closed_ = false;
+  _TryCatch* try_catch_ = nullptr;
+
+  // This is only for backwards compatibility, semantically it's always empty.
+  v8::Local<v8::Value> failure_return_value_;
+};
+
+CallbackScope::CallbackScope(Isolate* isolate,
+                             Local<Object> object,
+                             async_context asyncContext)
+  : private_(new InternalCallbackScope(Environment::GetCurrent(isolate),
+                                       object,
+                                       asyncContext)) {
+  private_->UseTryCatch();
+}
+
+CallbackScope::~CallbackScope() {
+  delete private_;
+}
+
+InternalCallbackScope::InternalCallbackScope(Environment* env,
+                                             Local<Value> object,
+                                             const async_context& asyncContext)
+  : env_(env),
+    async_context_(asyncContext),
+    callback_scope_(env) {
+  CHECK(!object.IsEmpty());
+  if (object->IsObject())
+    object_ = object.As<Object>();
+
+  // If you hit this assertion, you forgot to enter the v8::Context first.
+  CHECK_EQ(env->context(), env->isolate()->GetCurrentContext());
+
   if (env->using_domains()) {
-    CHECK(recv->IsObject());
-    disposed_domain = DomainEnter(env, object);
-    if (disposed_domain) return Undefined(env->isolate());
+    CHECK(object->IsObject());
+    failed_ = DomainEnter(env, object_);
```

Comment:
But, like `MakeCallback`, this has a lot more failure modes than just trying to enter a disposed domain? There are other cases in which `failed_` is set and in which `MakeCallback` previously bailed out.

Reply:
Like, if you have suggestions that make sense across the board, I'm all for it. Or we could split this into different boolean values, but that might unnecessarily increase complexity.

Reply:
what are the states? would a simple bit mask with a couple flags suffice?

Reply:
@jasnell Currently and previously, those are:

- Domain enter/exit callback was acting on a disposed domain
- The invoked callback itself threw
- Calling process.nextTick() afterwards threw

> would a simple bit mask with a couple flags suffice?

@jasnell Just to be clear, we would be setting different bits in a bit mask, but would never inspect those values, only care whether one was set. Am I understanding correctly that that's what you're suggesting?

Reply:
So `failed_` handles all failure cases. I see.

The reason the domain callbacks wrap the async wrap callbacks is that if the domain is disposed then `MakeCallback()` can return immediately. There's no cleanup that needs to be performed. Though I guess because of the way the code is now structured that fact does need to propagate.